Hello, AppEngine, TCell, and Java11

zyme opened the pull request:
These changes enable deployment of the backend to GAE.  Included here are DDP-4754  and DDP-4754.  One thing to get familiar with is GCP's [secret manager](https://cloud.google.com/secret-manager) (SM), which is proving to be way more fun than vault.  Versions!  UI!  Good CLI tools!  I took the opportunity to replace a few vault interactions with SM as part of GAE.

Along the way, expanded logging on the backend to include real IP and user email, as captured from JWT.

EDIT: Also upgraded to Java11.

SimoneMaiwald committed:
DDP-3730: - removing realm after logout
- adding user feedback when saving a default filter

SimoneMaiwald committed:
DDP-3730:
- fixing filterRoute from querying wrong list

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-3730:
- fixed npe

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-3730: - changed color of additionalMessage

SimoneMaiwald committed:
DDP-3730:
- fixed or search for activity data

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-3730:
- removed logging

Signed-off-by: simone <simone@broadinstitute.org>

pegahtah committed:
Merge pull request #16 from broadinstitute/DDP-3730_fixes

DDP-3730

pegahtah committed:
Merge pull request #17 from broadinstitute/DDP-3730_TissueViewPostProd

DDP-3730_TissueViewPostProd

SimoneMaiwald committed:
fixed typo

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
fixed tests

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-4637: - added a tool to set all kits of a ddp to be skipped for reminder events
- removed logging

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-4637: - NPE fix

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
DDP-4642: - added test for es search bar bug

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
Merge pull request #19 from broadinstitute/DDP-4637_reminderEmails

DDP-4637: …

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
adding rotated truststores

Signed-off-by: simone <simone@broadinstitute.org>

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
split repo

SimoneMaiwald committed:
updating gitignore

SimoneMaiwald committed:
Merge remote-tracking branch 'origin/develop' into DDP-4642_ESsearchBar

SimoneMaiwald committed:
DDP-4642: fixed search bug

SimoneMaiwald committed:
DDP-4642: removed debug logging

SimoneMaiwald committed:
Merge pull request #21 from broadinstitute/DDP-4642_ESsearchBar

DDP-4642

SimoneMaiwald committed:
DDP-4661: fixed filtering of multiple survey statuses

SimoneMaiwald committed:
Merge pull request #22 from broadinstitute/DDP-4661_surveyStatus

DDP-4661: fixed filtering of multiple survey statuses

SimoneMaiwald committed:
DDP-4672: fixed ES date not-empty search

SimoneMaiwald committed:
Merge pull request #23 from broadinstitute/DDP-4672_ESdateNotNull

DDP-4672: fixed ES date not-empty search

SimoneMaiwald committed:
DDP-4677: added new sub field

SimoneMaiwald committed:
Merge pull request #24 from broadinstitute/DDP-4677_newSubField

DDP-4677: added new sub field

SimoneMaiwald committed:
DDP-4692: fixed search for multiple data status

SimoneMaiwald committed:
DDP-4692: removed unused code

SimoneMaiwald committed:
Merge pull request #25 from broadinstitute/DDP-4692_dataStatus

Ddp 4692 data status

SimoneMaiwald committed:
DDP-4631: - adding contact

SimoneMaiwald committed:
Merge pull request #26 from broadinstitute/DDP-3496_mailingList

DDP-4631: - adding contact

SimoneMaiwald committed:
DDP-4712: - reducing call to ES

SimoneMaiwald committed:
Merge pull request #27 from broadinstitute/DDP-4712_refactor

DDP-4712: - reducing call to ES

SimoneMaiwald committed:
DDP-4712: - reducing call to ES

SimoneMaiwald committed:
Merge pull request #28 from broadinstitute/DDP-4712_refactorEScalls

DDP-4712: - reducing call to ES

SimoneMaiwald committed:
DDP-4712: - adding some debug logging

SimoneMaiwald committed:
DDP-4717: - adding display name for instances

pegahtah committed:
making changes for fixing the bug

SimoneMaiwald committed:
Merge pull request #30 from broadinstitute/DDP-4717_displayName

DDP-4717: - adding display name for instances

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-4695-numberFilter

SimoneMaiwald committed:
removing old mbc tests

pegahtah committed:
DDP-4695 making changes for fixing the bug

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-4695-numberFilter

pegahtah committed:
Merge pull request #31 from broadinstitute/DDP-4695-numberFilter

Merge remote-tracking branch 'origin/develop' into DDP-4695-numberFilter

SimoneMaiwald committed:
DDP-4658: - started to add ability to search for invitations data in ES

SimoneMaiwald committed:
Merge remote-tracking branch 'origin/develop' into DDP-4658_preferredLanguage

SimoneMaiwald committed:
DDP-4658: - search for invitation

SimoneMaiwald committed:
DDP-4658: - added preferredLanguage to KitRequestShipping object

SimoneMaiwald committed:
Merge pull request #32 from broadinstitute/DDP-4658_preferredLanguage

DDP-4658: preferred language

SimoneMaiwald committed:
DDP-4658: - fixed search bug

zyme committed:
WIP initial setup for running in appengine

zyme committed:
Brought over IP override from study-server

zyme commented on pom.xml:
These are necessary for using cloudsql  db connections.

zyme commented on pom.xml:
Needed an upgrade here for `JettyConfig` and getting the proper source IP into logs.

zyme commented on pom.xml:
At irregular intervals, `gcloud` was complaining about the size of the jar.  So instead of doing an uberjar, I broke this out into uploads of all dependent jars.

zyme commented on pom.xml:
All this  test coverage stuff wasn't doing anything, so I removed it.

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Although there is  already a `protected` version of this method on the superclass, I wanted to add the ability to support multiple URLs.  Browsers were giving me guff about using `*` and `included` token auth, so I addressed it here by echoing back the appropriate URL dynamically.

zyme commented on src/main/java/org/broadinstitute/dsm/jetty/JettyConfig.java:
Copy pasted from study-server.

zyme commented on src/main/resources/log4j.xml:
Logs should go to stdout so that they are picked up by appsec via stackdriver.

yufengwng commented on appengine/deploy/StudyManager.yaml:
If we're doing manual scaling, we should definitely implement GAE's lifecycle call `/_ah/start`, otherwise GAE will continuously reboot the instance. Since `DSMServer` is running a Spark server, it should be returning `404` right now, which GAE's documentation says is fine. But I'll recommend adding a route and returning a `200`.

yufengwng commented on appengine/deploy/StudyManager.yaml:
Did we mean `-Xms`?

yufengwng commented on pom.xml:
I'm a bit confused by this `exclusion`...

yufengwng commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
I recall @MocanaAtBroad added some magic header to facilitate better caching, something about `Access-Control-Max-Age`. Should we pull that over as well?

MocanaAtBroad commented on appengine/deploy/build-and-deploy.sh:
Consider giving dsm jar a nicer name.

MocanaAtBroad commented on appengine/deploy/build-and-deploy.sh:
vault integration via files? 😸

MocanaAtBroad commented on appengine/deploy/StudyManager.yaml:
Implementing /_ah/start is a good idea. Don't think it is as important for manual scaling as there will be no up/downscaling, but it is not hard to do. Check Yufengs implementation; wait to respond to /_ah/start request until app is completely initialized.
/_ah/stop too if there is cleanup needed when shutting down.

MocanaAtBroad commented on appengine/deploy/StudyManager.yaml:
Also consider defining a entry point for the jar so class name is not needed in command line.

MocanaAtBroad commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
You sure about hard-coding these?

MocanaAtBroad commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
https://github.com/broadinstitute/ddp-study-server/blob/6579631f8ccdc3cd480e621999a7fc12e94b889e/pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java#L602

Best bang for the buck performance boost. Allows browser to cache OPTION call result instead of preceding every API call with OPTION.

MocanaAtBroad commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Did you update the JDBC URL to use the GCP options?

MocanaAtBroad commented on pom.xml:
Just checked and there is finally a version 8.0.20 out that supposedly deal with time conversions properly (as you might recall we had to set timezones in the URL and in code to make sure we always were doing UTC). I suspect this is less of an issue in DSM.

MocanaAtBroad commented on pom.xml:
You probably dealt with it, but can't see the DB url so will mention it. I asked below about updating JDBC URL. Using their socket factory is one of the options in the URL. I believe same effect can be obtained when creating connection in code.

zyme commented on appengine/deploy/build-and-deploy.sh:
Vault?   What's  that?  :)

Check  out the  [secrets here](https://console.cloud.google.com/security/secret-manager?folder=&organizationId=&project=broad-ddp-dev)

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
@MocanaAtBroad to  answer  your question: yes, I'm sure that hardcoding the origin values  is a bad idea  :)

zyme commented on pom.xml:
Yep, I used the cloudsqlproxy-style URLs.  You can check  the secrets [here](https://console.cloud.google.com/security/secret-manager?folder=&organizationId=&project=broad-ddp-dev).   in  a UI!  With secret versions!  And no weird "don't forget `.data` and `jq` stuff"...

MocanaAtBroad commented on pom.xml:
I followed GAE's recommendation and setup the URL like below.
Requires granting permissions. If the proxy thing works then maybe its all good. But there might or might not be a performance penalty.
jdbc:mysql:///pepperdev?cloudSqlInstance=broad-ddp-dev:us-central1:ddp-dev&socketFactory=com.google.cloud.sql.mysql.SocketFactory&user=XXXX&password=YYYYY@&characterEncoding=UTF-8&useLegacyDatetimeCode=false&serverTimezone=UTC&useSSL=false&sessionVariables=innodb_strict_mode=on,tx_isolation='READ-COMMITTED',sql_mode='TRADITIONAL'

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Going with cloudsqlproxy, so we don't need any of the cert stuff here.

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Cloudsql proxy--no need for managing certs on the jdbc side.

zyme commented on pom.xml:
Ooops, this should have been an exclusion of `com.sparkjava` from `lddp`....

zyme committed:
PR  feedback

zyme commented on pom.xml:
Trying this out locally and it's causing some explosions with use of `beforeFirst()`.  Probably worth fixing and doing the upgrade...

```
2020-06-25 10:33:14,981 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.util.KitUtil.findSpecialBehaviorKits:530) - Checking pt N7TDWDSFIFV36KVADJJN for special behavior
2020-06-25 10:33:14,981 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.util.KitUtil.findSpecialBehaviorKits:530) - Checking pt PUI7OT7LTDRLBP18STRC for special behavior
2020-06-25 10:33:14,981 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.util.KitUtil.findSpecialBehaviorKits:530) - Checking pt 9C8LU8KT3OSBGTIPD0OW for special behavior
2020-06-25 10:33:14,982 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.util.KitUtil.findSpecialBehaviorKits:530) - Checking pt SDJZ2R73TGBHWGGNRCJ5 for special behavior
2020-06-25 10:33:15,020 INFO  [JobScheduler_Worker-1] (org.broadinstitute.ddp.db.TransactionWrapper.inTransaction:146) - retrieved connection...
2020-06-25 10:33:15,178 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.db.KitRequestShipping.getAllKitRequestsByRealm:390) - Collecting kit information
2020-06-25 10:33:15,217 INFO  [JobScheduler_Worker-1] (org.broadinstitute.ddp.db.TransactionWrapper.inTransaction:146) - retrieved connection...
2020-06-25 10:33:15,375 INFO  [JobScheduler_Worker-1] (org.broadinstitute.dsm.db.KitRequestShipping.getAllKitRequestsByRealm:470) - Got 0 participants kits in DSM DB for Osteo
2020-06-25 10:33:15,414 INFO  [JobScheduler_Worker-1] (org.broadinstitute.ddp.db.TransactionWrapper.inTransaction:146) - retrieved connection...
2020-06-25 10:33:15,572 ERROR [JobScheduler_Worker-1] (org.broadinstitute.dsm.jobs.DDPRequestJob.execute:58) - Some error occurred while doing pdf audit trail 
java.lang.RuntimeException: Error getting bookmark pdf_audit_kit
        at org.broadinstitute.dsm.util.DBUtil.getBookmark(DBUtil.java:33)
        at org.broadinstitute.dsm.util.PDFAudit.checkAndSavePDF(PDFAudit.java:18)
        at org.broadinstitute.dsm.jobs.DDPRequestJob.execute(DDPRequestJob.java:55)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.RuntimeException: Error getting bookmark pdf_audit_kit
        at org.broadinstitute.dsm.util.DBUtil.getBookmark(DBUtil.java:53)
        at org.broadinstitute.dsm.util.DBUtil.lambda$getBookmark$0(DBUtil.java:24)
        at org.broadinstitute.ddp.db.TransactionWrapper.inTransaction(TransactionWrapper.java:150)
        at org.broadinstitute.ddp.db.TransactionWrapper.inTransaction(TransactionWrapper.java:139)
        at org.broadinstitute.dsm.util.DBUtil.getBookmark(DBUtil.java:21)
        ... 4 more
Caused by: java.sql.SQLException: Operation not allowed for a result set of type ResultSet.TYPE_FORWARD_ONLY.
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.result.ResultSetImpl.last(ResultSetImpl.java:1711)
        at org.apache.commons.dbcp2.DelegatingResultSet.last(DelegatingResultSet.java:407)
        at org.apache.commons.dbcp2.DelegatingResultSet.last(DelegatingResultSet.java:407)
        at org.broadinstitute.dsm.util.DBUtil.getBookmark(DBUtil.java:43)
        ... 8 more
```

zyme committed:
PR feedback.  Upgrade to mysql-connector-java 8.0.20 seemed to cause a bunch of `Operation not allowed for a result set of type ResultSet.TYPE_FORWARD_ONLY` when resetting cursor position, which should be fixed by passing `ResultSet.TYPE_SCROLL_SENSITIVE,ResultSet.CONCUR_READ_ONLY` to statement creation.

zyme commented on src/main/java/org/broadinstitute/dsm/route/AssignParticipantRoute.java:
@MocanaAtBroad the pattern seems to be that with `8.0.20`, we have to set cursor modes explicitly for places that are going backwards through cursors.  I did a find/replace for areas of `rs.beforeFirst()` to try to find these.

MocanaAtBroad commented on src/main/java/org/broadinstitute/dsm/route/AssignParticipantRoute.java:
Had no idea that one could do these tricks with result sets! (though this looks like someone really really does not want to write more SQL queries than absolutely necessary!)

zyme committed:
PR feedback.  Added `ah/startup` route that will wait for everything (including liquibase) to finish before returning.

zyme committed:
PR feedback.  Added `ah/startup` route that will wait for everything (including liquibase) to finish before returning.

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
@MocanaAtBroad @yufengwng this turned out to be a bit more complicated than I originally thought, but it brings up something that I think we need to address in study-server as well: letting an instance start fully (allowing for liquibase to complete, which can take a while) before GAE sends requests to the instance.  Based on [this doc](https://cloud.google.com/appengine/docs/standard/java/how-instances-are-managed#startup), I think the approach here (to have `ah/start` block until things are ready) should work.  Otherwise, the default `404`  that spark returns will tell GAE "everything is fine--please start sending traffic to this instance", which is probably premature, since even the shortest deployments seem to take ~10s, without liquibase doing much.

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Also, I'm really glad y'all mentioned the whole `/ah/start` thing.  It was driving me kinda bonkers.

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
A sampling of more recent deploys shows cleaner `ah/start` calls and their duration.  Note the lack of `404`, `500`, or `503` at top.
<img width="741" alt="Screen Shot 2020-06-25 at 4 01 51 PM" src="https://user-images.githubusercontent.com/1653048/85789809-3aefda00-b6fd-11ea-94c7-23d031aa0ec7.png">

yufengwng commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Does this `setupJetty()` call need to go before `registerAppEngineStartupCallback()`? Mounting any routes or filters will have initialized the Spark server, which is what is being done in `registerAppEngineStartupCallback()` and `setupRouting()`?

zyme commented on pom.xml:
Gaah, I totally forgot the `UTC` url params

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Yep!

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
Looking at the  DSM logs (with the real IP), it doesn't look like the order matters here.  This  order of calls still produces  the correct IP address in DSM logs.

zyme committed:
Expanded logging to include IP and username in backend logs.

zyme commented on src/main/java/org/broadinstitute/dsm/route/LoggingFilter.java:
Was surprised  that we  aren't logging IP and JWT user id info--so I added it, since this is an important part of compliance.

<img width="1184" alt="Screen Shot 2020-06-29 at 11 46 30 AM" src="https://user-images.githubusercontent.com/1653048/86027075-3ded1200-b9fe-11ea-8af8-a4764750271e.png">

yufengwng commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
I still think it's a bit risky. From my reading of Spark's internals, the embedded server is initialized in a separate thread, so I suspect we might just be getting lucky that our custom factory is getting added before the thread goes through it's paces.

zyme committed:
WIP  with jdk11

zyme committed:
WIP  with jdk11

zyme committed:
WIP  with jdk11

zyme committed:
Added comment for avoiding superclass's setupRouting()

zyme committed:
Updated readme for java11

zyme committed:
Merge branch 'appengine-jdk11' into appengine

# Conflicts:
#	src/main/java/org/broadinstitute/dsm/DSMServer.java

zyme commented on pom.xml:
Woot, java11!

zyme commented on src/main/java/org/broadinstitute/dsm/DSMServer.java:
@pegahtah this change fixed the classnotfoundexception stuff.  Can you try this locally?

MocanaAtBroad commented on pom.xml:
I don't think you need sqlproxy to run in GAE. Just need to grant permissions to the service. Read somewhere (don't have the reference, sorry) that sqlproxy introduced latency.

yufengwng commented on pom.xml:
Looks like a repeat of the one on L15.

yufengwng reviewed (commented):
Overall looks good to me. Somehow I don't have the "Resolve conversation" button when viewing this PR in GitHub. Please feel free to close out comments.

zyme commented on pom.xml:
Sorry for the confusion--when I said "cloudsqlproxy-style", I meant the urls like `cloudSqlInstance=broad-ddp-dev:us-central1:ddp-dev`.  So not the cloudsqlproxy per se, but urls that reference gcp project, zone, and cloudsql instance name instead of IP.

zyme commented on pom.xml:
Double import, just to make double sure 🤦

zyme committed:
PR feedback

zyme committed:
Muted JWT error logging since it's just noise for unprotected routes.  Removed auth0 imports since they were causing issues with lddp infrastructure.

zyme committed:
Fixing bug in filter saving

zyme committed:
Should always use --stop-previous-version when deploying.  Updated PDF processing so that separate credentials aren't required when using GAE, which required updating lddp-backend.  Fixed app.yaml because order of args disabled tcell.

zyme committed:
Wrapping up appengine setup.  Added details to README.

zyme committed:
Moving cors config out to a config file

pegahtah committed:
Ddp 4769 batch kit endpoint (#35)

* DDP-4769 created an endpoint to get kit status for a batch of participants

* small renaming

* * changes made for handling the case that the pt is doesn't exist in dsm

* changed path to `/app/`

* Changed the JSON structure

* Changed the output as a `KitStatus`. Changed the class of `KitStatus` to have the `ddpKitRequestId`.

zyme committed:
Updated docs to include mvn package manager in settings.xml

zyme committed:
Merge branch 'appengine' of github.com:broadinstitute/ddp-study-manager into appengine

zyme committed:
Updated docs to include mvn package manager in settings.xml

MocanaAtBroad commented on appengine/deploy/build-and-deploy.sh:
One thing you might want to consider here is deleting older versions of deployed service. There is an overall limit in in the GCP project. On the Pepper side, right after we deploy, we try to remove any older versions. 
What service count look like right now in ddp-dev:
<img width="254" alt="Screen Shot 2020-07-23 at 7 11 41 AM" src="https://user-images.githubusercontent.com/29984380/88280521-f469ad00-ccb3-11ea-9a82-aae9af1dbcf3.png">

How we deal with it in CircleCi:
https://github.com/broadinstitute/ddp-study-server/blob/7c7843b080b0fd9289178c38ea9477025cc99a18/.circleci/config.yml#L294

Removing versions sometime simply fails. If that happens we just let it go and try to remove it again next time there is a deployment.

pegahtah committed:
batchkits-fix (#36)

* Changed where I set the route

* changing the BatchKitsRoute for an empty input

* changing the origin header since it is not being set by the browser sometimes

zyme committed:
DDP-4815 first draft of placing orders with CareEvolve for COVID-19
* added various json objects that match CareEvolve's examples
* updated README with some tidbits for setting up local dev
* added script for reading secrets for local development

zyme committed:
DDP-4815 WIP trying to get participant data for testboston

zyme committed:
DDP-4815 WIP fixed filter so it returns a participant

zyme committed:
DDP-4815 running real queries against ES and translating DDP data into CareEvolve

zyme committed:
DDP-4815 running real queries against ES and translating DDP data into CareEvolve

zyme committed:
DDP-4815 temporarily disabling test

zyme committed:
DDP-4815 temporarily disabling test

zyme committed:
DDP-4815 PR feedback
* Added collection time
* Fixed camel casing
* Added missing fields

zyme committed:
DDP-4815 PR feedback

yufengwng committed:
DDP-4848 enable login via SAML (#39)

yufengwng committed:
DDP-4900 use a proxy for elasticsearch client (#41)

zyme committed:
Merge branch 'appengine' into DDP-4815_CareEvolveOrder

zyme committed:
DDP-4815 capturing collection date/time

zyme committed:
DDP-4815 added  mappings for  race, ethnicity, and sex

zyme committed:
DDP-4815 cleanup

zyme committed:
DDP-4815 if multiple races are selected, use "other".

pegahtah committed:
changing the B8 instance and the `/api/` to `/ddp/` (#43)

zyme committed:
DDP-4815  PR feedback

zyme committed:
DDP-4815 added param for pickup time so  that we can use this in case the participant hasn't entered collection time in the symptom activity.

zyme committed:
Merge pull request #38 from broadinstitute/DDP-4815_CareEvolveOrder

DDP-4815 Placing orders with CareEvolve for COVID-19

pegahtah committed:
Appengine testboston gbf (#40)

* DDP-4629: - adding subkits to automated kit requests

Signed-off-by: simone <simone@broadinstitute.org>

* DDP-4630: - added needsApproval to KitDetail
- fixed Unit Tests

Signed-off-by: simone <simone@broadinstitute.org>

* DDP-4629: removed unused import

* DDP-4630: moved class

* DDP-4630: added needsApproval to ddp_kit

* DDP-4631: - added query for kits waiting for approval
- changed query for kits without label
- added new user roles

* DDP-4631: - adding approval and denial logic
- renaming and adding columns

* * Changes made to test units made previously for Promise to use the same ones for testboston.
* pom.xml got changed for `GBFRequestUtil.java` methods to work in Java 11 as well.

* pom changes

* Changes in POM and test classes to make the tests run

* * connecting GBF as an external shipper to DSM.

* GBF changes to:
* notify pepper of GBF shipped kits
* insert external order numbers for new orders
* insert the job in the DSM Server
* alert with errors if kits are not ordered or are cancelled

* More logging

* Cleaning code up

* only changing the status and time update IF kit status has changed, to know when it was updated

* Only change status when it has changed, so the update time is real not based on the cron job

* External Shipper job set up

* loggin an error when unable to trigger DDP

* logging stuff from gbf for debugging

* changed logging

* Confirmation XML s are now  being processed.

* * UPS tracking json objects

* * new Get request method with custom headers

* * Getting UPS response with password now
* added mocks for sending pepper received and delivered notifications -> gonna get removed in the next version

* * updating scan_date when gets SHIPPED

* * final changes to the confirmation XML

* * PR comment

* * removed unwanted todos

* * updating ordering and pr comments

* * Pr Comments

Co-authored-by: simone <simone@broadinstitute.org>

pegahtah committed:
Ups integration (#46)

* DDP-4815 made a bit more robust when some activity data is missing

* * Added UPS Tracking integration

* * new columns for keeping status

* * changes for getting ddpInstance with InstanceRole

* * debug changes

* * updating scan_date when gets SHIPPED

* * updating scan_date when gets SHIPPED

* * updating using DELIVERY as  a variable

* * getting a list of instances

* * getting tracking or return number now

* * updating RECEIVED and DELIVERED time

* * fixed typo in receive query

* * updated the number of results expected

* * getting short id from table

* * Gettiing a Mao instead of a set to avoid duplicates

* * fixed duplicates

* * null shipping

* * PR Comments

* * pr comments

* * pr comment

* * getting time of in transit as well

* * Found Bug

* * Added order by to avoid sending the wrong label to CE

* * add check for ordered or not
* pr comments

* * making sure the right kit label  is in the map

Co-authored-by: Andrew Zimmer <andrew@broadinstitute.org>

pegahtah committed:
testResults-pubsub-integrartion (#42)

* Liquibase update file for keeping test results

* Changed where I set the route

* Making the classes and models, triggers, jobs

* Job class

* Adding a column to `ddp_kit` to store the results and to log them into it.

* reading test results from db to show on participant list

* Changed subscriber

* The result Object updated

* * Changed the test result object to what Birch returns
* Telling Study Server about the new result
* Removed the timeout mechanism from pubsub receiver

* triggering ddp with the testResult Object

* * Applied PR comment
* result is saved as a JSON array
* duplicate check

* * logging error for non matching not corrected result

* PR comments

* Changed where I set the route

* changes

* * Sending a `DSMTestResult` object to Pepper instead of `TestBostonResult`
* Connecting to a DSM owned pubsub topic instead of Birch owned

* * Sending a `DSMTestResult` object to Pepper instead of `TestBostonResult`
* Connecting to a DSM owned pubsub topic instead of Birch owned

* * pom changes to include google storage

* * logging

* * PR comments
* removed liquibase unused column
*  made changes for prod
* reading project id from vault

* removing sample ID and using external order id

* * logs

* * added versions to pom

* Trying to reconcile conflicting GCP lib versions

* removed result_date

* removed result_date

* Fixing version issues  with GCP transient deps

* Update src/main/java/org/broadinstitute/dsm/DSMServer.java

Co-authored-by: Yufeng Wang <yufeng@broadinstitute.org>

* * removed unused pubsub job

Co-authored-by: Andrew Zimmer <andrew@broadinstitute.org>
Co-authored-by: Yufeng Wang <yufeng@broadinstitute.org>

pegahtah committed:
* 2 for number of subkits added to getKitDDPNotification

pegahtah committed:
* fixed Bugs

pegahtah committed:
* to avoid future bugs I think we should notify pepper inside a transaction

pegahtah committed:
* added an empty method

pegahtah committed:
* add check for junk orders

pegahtah committed:
* changed select command for kits based on their label to avoid junks

pegahtah committed:
* adding email notifications (#47)

* * adding email notifications
* adding a fix for _1 junk orders on CE

* * pr comments

* * pr comments

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into prod_09232020 (#50)

* * changed `ldd-backend`

* * changing the received_time

* Pruned some things from pom (#48)

* Pruned some things from pom

* Removed pubsub since it should come in from lddp-backed

* * changed `ldd-backend`

* multiple fixes (#49)

* removing credential file for google buckets. not needed with gae

* changes for test boston ui fix

* - fixing sample report and summary for test boston
- changing pom to use new lddp version

* - changes quartz version suggested by Dependabot

Co-authored-by: pegahtah <ptaheri@broadinstitute.org>
Co-authored-by: SimoneMaiwald <simone@broadinstitute.org>

* * received date change

Co-authored-by: Andrew Zimmer <andrew@broadinstitute.org>
Co-authored-by: SimoneMaiwald <simone@broadinstitute.org>

SimoneMaiwald committed:
- added test results to UI response

SimoneMaiwald committed:
- added ups shipment status to participant list and page

pegahtah committed:
* commented the received date out

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into prod_100520

pegahtah committed:
* added xmls to pom

pegahtah committed:
* fixes in the UPS select query

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into prod_100520 (#51)

* * Single thread for pub sub

* * removed unnecessary todo

pegahtah committed:
NPE fix on prod (#52)

* * Single thread for pub sub

* * removed unnecessary todo

* * fix for npe error

* DDP-5095 added profile creation date as the stable sort order for paged query.

* * another fix on npe

* *

* DDP-5109 added retry loop when placing an order in CE (#54)

Co-authored-by: Andrew Zimmer <andrew@broadinstitute.org>

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into prod_100520 (#55)

* * Single thread for pub sub

* * removed unnecessary todo

* * fix for npe error

* DDP-5095 added profile creation date as the stable sort order for paged query.

* * another fix on npe

* *

* DDP-5109 added retry loop when placing an order in CE (#54)

Co-authored-by: Andrew Zimmer <andrew@broadinstitute.org>

pegahtah committed:
uplod kit for test boston (#53)

* * changes to have a phone number in kit upload object
* changes to remove the kit upload bug for duplicate kits
* writing subkits of an uploaded kit into db

* * adding external order number

* * external order number added for subkits

* * added the `KitUtil.setReceived`

* removed todo

* pr comments

* merge errors

* Using `setReceivedDate` for setting the `receivedDate`

* * caught phoneNumber null bug

SimoneMaiwald committed:
[TB] missing columns (#56)

* - added missing external order columns to participant list

* - adding kitType

SimoneMaiwald committed:
[TB] Fixes (#57)

* - added missing external order columns to participant list

* - adding parameter to change format of fields and hiding ES dsm columns for studies

* DDP-4934: only sending out emails for corrected and positive test results

* DDP-5075: - adding display setting to hide mr and tissue workflow fields
- adding new user access role to see participant list
- removed unused class

* DDP-5120: - getting preferred language from study-server

* getting preferred languages only for gen3 studies

* DDP-5126: -fixed invitation date search

* DDP-5129: -fixed rights to share/delete filters

* DDP-5159: - fixed filtering error

* added errors in case participant is not found in ES

pegahtah committed:
* added more logging for better debugging (#58)

* changed kit not shipped error to warning

pegahtah committed:
DDP-5173 (#59)

* * add check for complete fields in the address

* * placing CE orders in case of no updates between label created and deliver statuses.

* * placing an order when the new status is not LABEL_CREATED anymore

* * KitUploadRoute bug for different subkits with different ddp_kit_request_id fixed

pegahtah committed:
* only querying for not delivered kits

pegahtah committed:
* getting GBF from the beginning

pegahtah committed:
* logging added

zyme committed:
DDP-5323 When kit is in transit or delivered, trigger notifications

zyme committed:
Merge pull request #61 from broadinstitute/DDP-5323_prod_11022020

DDP-5323 Use in transit and delivered status to trigger events

pegahtah committed:
* changes in the update query

SimoneMaiwald committed:
filtering test results

zyme committed:
DDP-5323 fixing bug that was prematurely marking a kit as returned to broad and therefore locking the longitudinal activity before the kit was delivered to the participant.

zyme committed:
Merge pull request #63 from broadinstitute/DDP-5323_prod_11022020

DDP-5323 fixing premature read-only of tbos symptoms activity

zyme committed:
DDP-5345 Use UPS history for collection date (#64)

DDP-5345

SimoneMaiwald committed:
fixing in case language endpoint doesn't respond

zyme committed:
DDP-5353 GBF start time should be fixed, not dynamic.

pegahtah committed:
* adding new upload reason for kit upload

pegahtah committed:
*removed error

zyme committed:
DDP-5353 cron for GBF job does not need a sub-constraint separate from regular cron

pegahtah committed:
* only querying for not delivered kits

pegahtah committed:
* changes in `Event` class to stop using the ddp version and using `KitEvent` class instead.
*Added `ddpKitRequestedId` and `uploadReason` to `KitEvent`

pegahtah committed:
* added route for getting carriers for upload kits

pegahtah committed:
* PR comments applied

pegahtah committed:
* adding `NORMAL` to Kit Events that don't have an upload reason

pegahtah committed:
* `UPSTrackingJob` named changed to `TestBostonUPSTrackingJob`

zyme committed:
DDP-5409
Loosened up query be removing kit_label and external_response conditions, and imposing that test_result be null and that both outbound and returned packages cannot both be delivered.

pegahtah committed:
* fixed bug in updating return status

zyme committed:
Updating queries so that most recent kits are processed first, and to allow for single row updates even when we might expect two.

zyme committed:
Wrapping update calls in try/catch for better robustness

zyme committed:
Fixing query--forgot that not like returns null when dealing with nulls

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into UPSCodeUpdate

# Conflicts:
#	src/main/java/org/broadinstitute/dsm/jobs/ExternalShipperJob.java
#	src/main/java/org/broadinstitute/dsm/jobs/TestBostonUPSTrackingJob.java

pegahtah committed:
* `UPSTrackingJob` named changed to `TestBostonUPSTrackingJob`
*  Corrected a bug in TestBostonUPSTrackingJob
* getting `upload_reason` in queries

pegahtah committed:
* changed the ExternalShipperJob to start looking from 5 days ago

pegahtah committed:
* `LAST_KIT` replaced with `DDP_KIT_REQUEST_ID`

pegahtah committed:
* `Instant` being used for creating a dynamic start time

pegahtah committed:
* fixed Kit_EVENT error

zyme committed:
Wrap order confirmation in a big try/catch (#71)

pegahtah committed:
* Fixed KitLable logging
*Fixed access needed for `CarrierServiceRoute`

pegahtah committed:
* PR comments

SimoneMaiwald committed:
Merge pull request #72 from broadinstitute/prod_12142020

* Fixed KitLabel logging and access for `CarrierServiceRoute`

zyme committed:
DDP-5447 SQL was missing `,` (#73)

* DDP-5447 query was missing `,`
Co-authored-by: simone <simone@broadinstitute.org>

Irakli-Khaladze committed:
Start another approach for slack notification

Irakli-Khaladze committed:
Create SlackAppender class

gmakharat committed:
overriding methods for slack

Irakli-Khaladze committed:
Add payload to the message

gmakharat committed:
added try/catch to handle http request

Irakli-Khaladze committed:
Use slack appender in log4j file

gmakharat committed:
DDP-5446 | started creating test for slack notification

Irakli-Khaladze committed:
Continue creating test for slack notification

gmakharat committed:
continued working on test for slack appender

gmakharat committed:
finished test for slack appender

zyme committed:
DDP-5479 Reducing the number of calls we make to GBF status endpoint (#82)

DDP-5447 
* increasing memory based on experience with study-server
* query for one participant at a time in ES
DDP-5479 
* shuffling list of GBF data so that we make _some_ progress on different samples if we do crash
* addressing lockwait error by using same connection throughout.
* ignoring things that have been ordered in CE
* ignoring UPS status--it's not relevant here
* eliminating redundant calls for the same kit
* DDP-5478 inline query `tracking_id` to avoid ambiguous `kit_label`
* Instead of querying large lists of `KitRequest`s and then iterating through them, stream through the data, updating as we go.  Along the way, to fix some non-transactional behavior, it was necessary to add `Connection` param to many methods, which rippled into a number of classes as this is a fairly wide refactor in addition to the streaming change.
* DDP-5479 Reducing the number of calls we make to GBF status endpoint by:
* DDP-5480: pointing to the right role
Co-authored-by: simone <simone@broadinstitute.org>
Co-authored-by: pegahtah <ptaheri@broadinstitute.org>

zyme committed:
DDP-5479 removing minimum memory setting

SimoneMaiwald committed:
DDP-5479: adding heap dump tool
draft for participant filtering

SimoneMaiwald committed:
Merge branch 'appengine' of https://github.com/broadinstitute/ddp-study-manager into DDP-5479_heapDump

SimoneMaiwald committed:
changing the bucket name to fit all other dsm buckets

SimoneMaiwald committed:
more improvements for ParticipantWrapper

SimoneMaiwald committed:
added some logging

SimoneMaiwald committed:
Merge pull request #84 from broadinstitute/DDP-5479_heapDump

DDP-5479: java heap dump

SimoneMaiwald committed:
adding a check to not update old data

SimoneMaiwald committed:
changed check to include CE_order

SimoneMaiwald committed:
DDP-5496

SimoneMaiwald committed:
DDP-5496: removing CE_order from update query

SimoneMaiwald committed:
Merge pull request #85 from broadinstitute/fixTrackingBug

adding a check to not update data of other kit

SimoneMaiwald committed:
DDP-5479: - changing response

SimoneMaiwald committed:
DDP-5507: - added hotfix for ESC

SimoneMaiwald committed:
Merge remote-tracking branch 'origin/appengine' into DDP-5479_response

SimoneMaiwald committed:
DDP-5479: - returning response per OutputStreamWriter

SimoneMaiwald committed:
Merge pull request #88 from broadinstitute/DDP-5479_response

DDP-5479: changing response

SimoneMaiwald committed:
DDP-5479: refactoring ups job (#86)

* DDP-5476: removing unused imports

* DDP-5479: - refactored to streaming
- changed update sql to contain external order number

* DDP-5479: - added external order number to log info

* DDP-5479: - refactoring

* DDP-5479: - changed update query to have ddp_kit_request table in where condition

* DDP-5479: - adding indexing for tracking is columns

* DDP-5479: - changing kit label into separate method

* * change selection of `KitDDPNotification` to be by external order number instead of tracking id to make sure we get 2.

* * fixing NPE when querying for kits that don't have tracking id

* DDP-5479: - adding better logging for failed event triggering at EventUtil

Co-authored-by: pegahtah <ptaheri@broadinstitute.org>

pegahtah committed:
* NPE error fixed (#89)

SimoneMaiwald committed:
DDP-5515: - fix for ESC

SimoneMaiwald committed:
Merge pull request #91 from broadinstitute/DDP-5515_filterESC

DDP-5515: - fix for ESC

pegahtah committed:
GBFStatus endpoint bad format fix (#92)

* The logs show we never entered line 211 and we are querying the endpoint wrong.

* * logging the error

Irakli-Khaladze committed:
Merged appengine updates

Irakli-Khaladze committed:
Merge branch 'DDP-5446-v2_adding_slack_api' of https://github.com/broadinstitute/ddp-study-manager into DDP-5446-v2_adding_slack_api

Irakli-Khaladze committed:
Get config parameters during setup

gmakharat committed:
fixed test of slackappender

Irakli-Khaladze committed:
Change system.out with logger

Irakli-Khaladze committed:
Use JOB_DELAY and NON_JOB_DELAY fields where neccessary

Irakli-Khaladze committed:
Some cosmetic changes

gmakharat committed:
removed unnecessary ignores from gitignore

pegahtah committed:
Merge remote-tracking branch 'origin/filterResult' into DDP-5532_filterChanges

gmakharat committed:
changed username from DSM to Study-Manager

pegahtah committed:
* Type `JSON_ARRAY` has been added to the automata for parsing filtering so it  can not be saved and  re applied

gmakharat committed:
Merge pull request #94 from broadinstitute/DDP-5446-v2_adding_slack_api

Changing functionality to send error notification to slack channel

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into DDP-5532_filterChanges

gmakharat committed:
DDP-5446 | added more exception information to slack message

SimoneMaiwald committed:
DDP-5556: - flagging second initial_biopsy as deleted

gmakharat committed:
Merge pull request #96 from broadinstitute/DDP-5446-v2_adding_slack_api

DDP-5446 | added more exception information to slack message

pegahtah committed:
Merge branch 'DDP-5532_filterChanges' into DDP-5441_UploadCheck

SimoneMaiwald committed:
DDP-5556: - add deleted flag to missed sql

SimoneMaiwald committed:
Merge remote-tracking branch 'origin/appengine' into DDP-5556

SimoneMaiwald committed:
Merge pull request #97 from broadinstitute/DDP-5556

DDP-5556: Duplicate initial biopsy

pegahtah committed:
DDP-5532_filterChanges (#95)

* filtering test results

* fixing in case language endpoint doesn't respond

* * Type `JSON_ARRAY` has been added to the automata for parsing filtering so it  can not be saved and  re applied

* * refactoring

Co-authored-by: simone <simone@broadinstitute.org>

pegahtah committed:
DDP-5532_filterChanges (#98)

* filtering test results

* fixing in case language endpoint doesn't respond

* * Type `JSON_ARRAY` has been added to the automata for parsing filtering so it  can not be saved and  re applied

* * refactoring

* * format  fix

Co-authored-by: simone <simone@broadinstitute.org>

pegahtah committed:
Merge remote-tracking branch 'origin/appengine' into DDP-5441_UploadCheck

# Conflicts:
#	src/main/java/org/broadinstitute/dsm/db/ViewFilter.java

pegahtah committed:
* added changes for checking backend participant exists and ids valid

pegahtah committed:
* fix for getting the ES data with large list. (#101)

SimoneMaiwald committed:
DDP-5076: adding more information (#100)

* DDP-5076: - adding proxy information
- adding new activity question types

* DDP-5076: - rollback of change (needs more testing!)

* DDP-5076: - rollback of change

SimoneMaiwald committed:
DDP-4702: - dynamic form for AT
- Kit search for AT

zyme committed:
DDP-5633 Retry UPS queries a few times before failing (#103)

* DDP-5633 retry UPS queries a few times before giving up

* DDP-5633 interim failure should be `warn`, not `error`

Irakli-Khaladze committed:
DDP 5235 first version of API for getting messages from client, sending to pubsub topic, getting result and sending back to client (#99)

* Add publisher for message dsm to dss and subscriber for dss to dsm, add websocket to handle this process

* Add functionality to save session and send message to it in subscriber

* added authentification for websocket using jwt token from url

* Delete commented unnecessary code

* Change structure of messages and fix several issues

* DDP-5235 | Add message table and routes for editing participant

* DDP-5235 | Add second route in route path and some minor changes

* Change http method to put from post

* changes to set status to sent back

* Minor changes to avoid possible bug

* DDP-5235 | Delete unnecessary line, which was for testing purposes

* Fix issue in DownloadPDFRoute

* Revert changes, which were enabling usage of sockets, fix typo

* DDP-5235 | Check if user has access

Co-authored-by: gmakhara <gmakhara@broadinstitute.org>

SimoneMaiwald committed:
Merge remote-tracking branch 'origin/appengine' into DDP-4702_ATform

Irakli-Khaladze committed:
DDP-5369 | add study_guid column in message table, some changes in inserting message

gmakharat committed:
DDP-5441 | changed check of kit upload participant from internal db participant to participant from elastic search, made code refactoring

Irakli-Khaladze committed:
DDP-5639 | Add study guid to table ddp_instance not message

gmakharat committed:
Merge branch 'appengine' of https://github.com/broadinstitute/ddp-study-manager into DDP-5441_UploadCheck

Irakli-Khaladze committed:
Rever EditParticipantMessage back to normal, add functionality to get study guid with instance name, minor change in subscriber to avoid less expected error

Irakli-Khaladze committed:
Small refactoring and checking new access role for editing participant

gmakharat committed:
DDP-5446 | changed visual of error message, extracted exceptions, which are thrown by the code written by developer, to include into slack message, added GCP link to redirect to error logs in google cloud platform

gmakharat committed:
DDP-5441 | DDP-5440  according to ticket DDP-5440, changed requirements, removed participantId from kit upload, instead of checking participantId, checks participant full name if the participant exists in the study

SimoneMaiwald committed:
DDP-4702: - dynamic form for AT
- Kit search for AT

SimoneMaiwald committed:
Merge pull request #102 from broadinstitute/DDP-4702_ATform

DDP-4702: - dynamic form for AT

gmakharat committed:
DDP-5446 | started adding feature for gcp log

Irakli-Khaladze committed:
Merge pull request #105 from broadinstitute/DDP-5639_message-study-guid

DDP-5639 message with study guid and DDP-5641 checking new role

gmakharat committed:
DDP-5441 | removed participant names from log, code refactored as suggested

gmakharat committed:
Merge branch 'appengine' of https://github.com/broadinstitute/ddp-study-manager into DDP-5446-v2_adding_slack_api

gmakharat committed:
DDP-5446 | added redirection url of gcp logs

Irakli-Khaladze committed:
DDP-5235 Edit participant test

gmakharat committed:
Merge pull request #106 from broadinstitute/DDP-5441_UploadCheck

DDP-5441 | kit upload check

SimoneMaiwald committed:
merging hotfix from 2/16/2021 into appengine (#109)

gmakharat committed:
Merge branch 'appengine' of https://github.com/broadinstitute/ddp-study-manager into DDP-5446-v2_adding_slack_api

gmakharat committed:
Merge pull request #108 from broadinstitute/DDP-5446-v2_adding_slack_api

DDP-5446 | changed visual of error message, extracted exceptions, whi…

SimoneMaiwald commented:
going to close old PRs now
